src/ObjCRuntime/TrampolineBlockBase.cs

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace ObjCRuntime {
+
+	public abstract class TrampolineBlockBase {
+			protected IntPtr blockPtr;
+
+			[DllImport ("/usr/lib/libobjc.dylib")]
+			static extern IntPtr _Block_copy (IntPtr ptr);
+
+			protected unsafe TrampolineBlockBase (BlockLiteral *block)
+			{
+				blockPtr = _Block_copy ((IntPtr) block);
+			}
+			
+			~TrampolineBlockBase ()
+			{
+				Runtime.ReleaseBlockOnMainThread (blockPtr);
+			}
+
+			protected unsafe static object GetExistingManagedDelegate (IntPtr block)
+			{
+				if (!BlockLiteral.IsManagedBlock (block))
+					return null;
+				return ((BlockLiteral *) block)->Target;
+			}
+	}
+}
```

Comment:
I think you can move the `invoker` field if you make the new class generic, and partially move the `Create` code as well:

```csharp
public abstract class TrampolineBlockBase<T> : where T: System.MulticastDelegate
{
    T invoker;
    protected T Invoker { get { return invoker; } }

    protected static bool Get<T> (IntPtr block, out T del) where T: System.MulticastDelegate
    {
        del = null;
        if (block == IntPtr.Zero)
            return true;
        if (BlockLiteral.IsManagedBlock (block))
            del = BlockLiteral.GetTarget<T> (block);
        return del != null;
    }
    // ...
}
```

and then the generated Create method becomes something like this:

```csharp
[Preserve (Conditional=true)]
[BindingImpl (BindingImplOptions.GeneratedCode | BindingImplOptions.Optimizable)]
public unsafe static global::AVFoundation.AVAssetImageGeneratorCompletionHandler Create (IntPtr block)
{
    if (Get<global::AVFoundation.AVAssetImageGeneratorCompletionHandler> (block, out var del))
        return del;
    return new NIDAVAssetImageGeneratorCompletionHandler ((BlockLiteral *) block).Invoke;
}
```

Reply:
Let me try (again), I tried something similar (w/generics) a while back (code has been sitting in a branch for a while) but I can't remember why I did not keep it

Reply:
hmm... I think it did not play well with other pieces I had for https://github.com/xamarin/xamarin-macios/issues/6792

Reply:
So while it [looks nicer](https://gist.github.com/spouliot/9130e0e3a9694d08f554290d08dc37f4) (not that many people should be reading this) it's actually a big bigger than the original (non generic) version.

generic

```
-rwxr-xr-x  1 poupou  staff  15868928 26 Sep 15:44 _ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/32bits/Xamarin.iOS.dll
-rwxr-xr-x  1 poupou  staff  12280832 26 Sep 15:44 _ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/mono/Xamarin.iOS/Xamarin.iOS.dll
-rwxr-xr-x  1 poupou  staff  17033216 26 Sep 15:44 _ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/64bits/Xamarin.iOS.dll
-rwxr-xr-x  1 poupou  staff  8365568 26 Sep 15:44 _ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/mono/Xamarin.TVOS/Xamarin.TVOS.dll
-rwxr-xr-x  1 poupou  staff  11623936 26 Sep 15:44 _ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/64bits/Xamarin.TVOS.dll
-rwxr-xr-x  1 poupou  staff  7275520 26 Sep 15:42 _ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/32bits/Xamarin.WatchOS.dll
-rwxr-xr-x  1 poupou  staff  5285888 26 Sep 15:43 _ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/mono/Xamarin.WatchOS/Xamarin.WatchOS.dll
lrwxr-xr-x  1 poupou  staff  38 26 Sep 10:11 _mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/git/lib/mono/Xamarin.Mac/Xamarin.Mac.dll -> ../../reference/mobile/Xamarin.Mac.dll
lrwxr-xr-x  1 poupou  staff  36 26 Sep 10:11 _mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/git/lib/mono/4.5/Xamarin.Mac.dll -> ../../reference/full/Xamarin.Mac.dll
-rwxr-xr-x  1 poupou  staff  20989952 26 Sep 15:44 _mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/git/lib/reference/full/Xamarin.Mac.dll
-rwxr-xr-x  1 poupou  staff  20993536 26 Sep 15:44 _mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/git/lib/reference/mobile/Xamarin.Mac.dll
-rwxr-xr-x  1 poupou  staff  20989952 26 Sep 15:44 _mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/git/lib/x86_64/full/Xamarin.Mac.dll
-rwxr-xr-x  1 poupou  staff  20993536 26 Sep 15:44 _mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/git/lib/x86_64/mobile/Xamarin.Mac.dll
```

I'll compare AOT sizes before committing - generic-based code surprised me more than once over the years :)

Reply:
ARM64 only, not linked, release w/LLVM backend

* original (before): 210,530,540 bytes
* non-generic version: 210,211,261 bytes
* generic version: 210,937,149 bytes

So once AOT'ed the generic version is actually bigger than the original/current.

Reply:
There's nothing like having real numbers!

Reply:
Always better than imaginary numbers I think :P